cpp/src/arrow/util/iterator.h

```diff
@@ -503,12 +521,12 @@ class FlattenIterator {
   explicit FlattenIterator(Iterator<Iterator<T>> it) : parent_(std::move(it)) {}
 
   Result<T> Next() {
-    if (child_ == IterationTraits<Iterator<T>>::End()) {
+    if (IsIterationEnd<Iterator<T>>(child_)) {
```

Comment:
It shouldn't be necessary to specify the template argument to `IsIterationEnd`
```suggestion
    if (IsIterationEnd(child_)) {
```

Reply:
Good catch.  Fixed.